Correção e melhorias rota CREATE & Error-handling rota CREATE

carmo-daniel opened the pull request:
Seguem as melhorias e correções recomendadas!

carmo-daniel committed:
Rota CREATE

carmo-daniel committed:
Validação da entrada e error-handling

carmo-daniel committed:
Correção de dupla requisição do modelo Jogo

pedro-de-oliveira-guedes committed:
Merge branch 'main' into main

pedro-de-oliveira-guedes reviewed (changes requested):
# Code Review
Opa, bom dia Daniel! Tudo tranquilo com você? 😄
Assim como da última vez, vou te dar umas ideias sobre como melhorar o código. Por favor, sinta-se livre para argumentar comigo e a equipe e definirmos como será feito o desenvolvimento. 😉

Vou dividir a revisão por arquivos e, como as rotas do usuário e dos jogos são bastante similares, farei a revisão de ambas juntas, ok? 🧐

---
## Arquivo ```Usuario.js```
Muito boa a sacada de tirar o ```{alter: true}```! Eu dei uma pesquisada aqui e ele realmente pode ser destrutivo para projetos em fase de produção. Não havia ocorrido nenhum erro para mim, mas no caso de alterações no arquivo de modelo da tabela, poderia prejudicar os dados já inseridos pelo usuário. 😮
Aí sim hein!? 😉

---
## Arquivo ```gameRoutes.js``` e ```userRoutes.js```
Parabéns Daniel, o arquivo das rotas tá muito bem escrito 🥳!
Algumas observações, porém, são válidas. Vou organizá-las por funcionalidade, tudo certo?
### Validação de entrada do usuário
Muito boa a validação dos dados utilizada Daniel! 😄
Mas ela pode ser melhorada em alguns aspectos. Por exemplo, será que a variável ```count``` é realmente necessária? 🤔
Poderiamos simplesmente mover o trecho de código:
```js
res.status(404).send('Some property sended in the body was not found');
```
para dentro de 
```js
if(!attributes[property]){
```

<blockquote>

Por favor, lembre-se de utilizar o comando ```return``` após devolver a resposta de erro para o cliente.
Isso evita o erro de enviar mais de uma resposta para uma única requisição. 😉

</blockquote>

Dessa forma, haverá menos ```ifs``` no código, ele ficará mais fácil de entender e poderá terminar mais cedo, caso alguma entrada do usuário seja inválida! 😮
### Criação do jogo/usuário na tabela
Com essa redução de ```ifs```, seria possível criar o jogo logo após validar a entrada do usuário. A seguir há um trecho de código mostrando melhor o que quero dizer. Não fiz as alterações que sugeri anteriormente, mas acho que dá pra ter uma ideia do que quero dizer. 😉
```js
for(let property in body){
    if(!attributes[property]){
        count++;
    }
}
await Jogos.create(jogo);
res.status(200).send(jogo);
```
Isso reduziria o tempo de execução do nosso algoritmo e pode também ser uma prática mais saudável para manutenção do código posteriormente, já que ele teria muito menos linhas e uma sintaxe mais direta. 🏆

## Arquivo ```gameRoutes.js```
Um comentário específico para este arquivo se fez necessário, isso devido ao trecho de código a seguir:
```js
if(count == 0 && jogo.IdUsuario!=null){
    await Jogos.create(jogo);
    res.status(200).send(jogo);
}
```
Por padrão, chaves estrangeiras conectando duas tabelas permitem valores nulos. Então, nesse caso, não seria necessário realizar esta verificação para a coluna ```IdUsuario``` em questão. Esta verificação poderia ser feita para outras colunas que foram especificamente modeladas para não aceitar valores nulos (vide arquivos de modelo), mas este é outro tipo de validação que creio não ser necessária para o projeto atual. 😉

Sendo assim, além da parte técnica que foi apresentada acima, temos também a interpretação do projeto, que indica que pode haver um problema em deixar este trecho de código em questão. 🤔
Se estamos modelando um sistema de uma locadora de jogos, é possível que, ao cadastrar um novo jogo no sistema, ele não esteja sendo alugado por ninguém. É inclusive muito provável que ele não esteja, já que ele acabou de ser cadastrado. 😃

Devido a essas questões, creio que seja melhor remover esta avaliação ao criar um novo jogo. Você não concorda? 🤔

---
## Conclusão
O código está muito bom Daniel, o progresso é notável! 😄
Parabéns, continue assim!!! 🚀🥳

carmo-daniel committed:
Melhorando eficiência e correção de bug

Reduzi as linhas de código para torná-lo mais eficiente e defini melhor o error handling para o arquivo de jogos.

pedro-de-oliveira-guedes committed:
Removendo ";" do final de validadeDeleteRequest.

pedro-de-oliveira-guedes reviewed (approved):
# Code Review
Chique demais da conta Daniel! 😉
Já vou incluir aqui! Muito obrigado pelas alterações! 😀
Parabéns mesmo pelo código, tá melhorando cada dia mais! 🥳🎉🏆
